elements/lisk-p2p/src/utils.ts

```diff
@@ -0,0 +1,224 @@
+	readonly bBytes: Buffer;
+	readonly cBytes: Buffer;
+	readonly dBytes: Buffer;
+}
+
+export const getIPBytes = (address: string): AddressBytes => {
+	const aBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	aBytes.writeUInt8(getIPGroup(address, 0), 0);
+	const bBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	bBytes.writeUInt8(getIPGroup(address, 1), 0);
+	const cBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	cBytes.writeUInt8(getIPGroup(address, 2), 0);
+	const dBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	dBytes.writeUInt8(getIPGroup(address, 3), 0);
+
+	return {
+		aBytes,
+		bBytes,
+		cBytes,
+		dBytes,
+	};
+};
+
+export const isPrivate = (address: string) =>
+	getIPGroup(address, 0) === 10 ||
+	(getIPGroup(address, 0) === 172 &&
+		(getIPGroup(address, 1) >= 16 || getIPGroup(address, 1) <= 31));
+
+export const isLocal = (address: string) =>
+	getIPGroup(address, 0) === 127 || getIPGroup(address, 0) === 0;
+/* tslint:enable no-magic-numbers */
+
+export const getNetwork = (address: string): NETWORK => {
+	if (isLocal(address)) {
+		return NETWORK.NET_LOCAL;
+	}
+
+	if (isPrivate(address)) {
+		return NETWORK.NET_PRIVATE;
+	}
+
+	if (isIPv4(address)) {
+		return NETWORK.NET_IPV4;
+	}
+
+	return NETWORK.NET_OTHER;
+};
+
+export const getNetgroup = (address: string, secret: number): number => {
+	const secretBytes = Buffer.alloc(SECRET_BUFFER_LENGTH);
+	secretBytes.writeUInt32BE(secret, 0);
+	const network = getNetwork(address);
+	const networkBytes = Buffer.alloc(NETWORK_BUFFER_LENGTH);
+	networkBytes.writeUInt8(network, 0);
+
+	// Get prefix bytes of ip address to bucket
+	const { aBytes, bBytes } = getIPBytes(address);
+
+	// Check if ip address is unsupported network type
+	if (network === NETWORK.NET_OTHER) {
+		throw Error('IP address is unsupported.');
+	}
+
+	// Seperate buckets for local and private addresses
+	if (network !== NETWORK.NET_IPV4) {
+		return cryptoHashSha256(
+			Buffer.concat([secretBytes, networkBytes]),
+		).readUInt32BE(0);
+	}
+
+	const netgroupBytes = Buffer.concat([
+		secretBytes,
+		networkBytes,
+		aBytes,
+		bBytes,
+	]);
+
+	return cryptoHashSha256(netgroupBytes).readUInt32BE(0);
+};
+
+// For new peer buckets, provide the source IP address from which the peer list was received
+export const getBucket = (options: {
+	readonly secret: number;
+	readonly targetAddress: string;
+	readonly sourceAddress?: string;
+}): number => {
+	const { secret, targetAddress, sourceAddress } = options;
+	const peerListType = sourceAddress ? NEW_PEERS : TRIED_PEERS;
+	const firstMod = peerListType === NEW_PEERS ? BYTES_16 : BYTES_4;
+	const secondMod = peerListType === NEW_PEERS ? BYTES_128 : BYTES_64;
+	const secretBytes = Buffer.alloc(SECRET_BUFFER_LENGTH);
+	secretBytes.writeUInt32BE(secret, 0);
+	const network = getNetwork(targetAddress);
+	const networkBytes = Buffer.alloc(NETWORK_BUFFER_LENGTH);
+	networkBytes.writeUInt8(network, 0);
+
+	// Get bytes of ip address to bucket
+	const {
+		aBytes: targetABytes,
```

Comment:
It would be nice to add a comment somewhere to explain what ABytes, BBytes, CBytes and DBytes mean to save the user having to dig through the code.